Subject: [PATCH] 02_UO_URTS Event Bus Channels Per Owner
---
Index: Assets/Scenes/Game.unity
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scenes/Game.unity b/Assets/Scenes/Game.unity
--- a/Assets/Scenes/Game.unity	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scenes/Game.unity	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -1308,6 +1308,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 11.722157
@@ -7462,6 +7466,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 8.530001
@@ -15379,6 +15387,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 5.87
@@ -15764,6 +15776,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 11.722157
@@ -15821,6 +15837,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 3.1378436
@@ -15952,6 +15972,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 14.5
@@ -17781,6 +17805,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 1038136032758847798, guid: d606f67b2169916478d3b1364e1b4428, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 1324784020702504064, guid: d606f67b2169916478d3b1364e1b4428, type: 3}
       propertyPath: m_LocalPosition.x
       value: -10.455597
@@ -18666,6 +18694,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 2.9078436
@@ -20110,6 +20142,10 @@
       propertyPath: m_Name
       value: Rifleman
       objectReference: {fileID: 0}
+    - target: {fileID: 5312076862503654571, guid: 89fa1dbd1317d914a9727e0841043e2c, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 5550838747094831301, guid: 89fa1dbd1317d914a9727e0841043e2c, type: 3}
       propertyPath: m_LocalPosition.x
       value: -8
@@ -22301,6 +22337,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 2.9078436
@@ -23323,6 +23363,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 6.1
@@ -25152,6 +25196,10 @@
     m_Modifications:
     - target: {fileID: 1422532545010619854, guid: 43943267dde60fe4a88cf5bd885c00e3, type: 3}
       propertyPath: m_Enabled
+      value: 1
+      objectReference: {fileID: 0}
+    - target: {fileID: 1422532545010619854, guid: 43943267dde60fe4a88cf5bd885c00e3, type: 3}
+      propertyPath: <Owner>k__BackingField
       value: 1
       objectReference: {fileID: 0}
     - target: {fileID: 1589499662120937400, guid: 43943267dde60fe4a88cf5bd885c00e3, type: 3}
@@ -28245,6 +28293,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 6.1
@@ -29760,6 +29812,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 14.25
@@ -30689,6 +30745,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 5.87
@@ -33335,6 +33395,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 3.1378436
@@ -33392,6 +33456,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 11.492157
@@ -35941,6 +36009,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 11.492157
@@ -36568,6 +36640,10 @@
     serializedVersion: 3
     m_TransformParent: {fileID: 0}
     m_Modifications:
+    - target: {fileID: 3843308947485064242, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
+      objectReference: {fileID: 0}
     - target: {fileID: 4078631732461238711, guid: 883b9682b823cb84cac42e4ad43d41d2, type: 3}
       propertyPath: m_LocalPosition.x
       value: 8.76
@@ -37130,6 +37206,10 @@
     - target: {fileID: 1076983579694789324, guid: 4d0c2e8c3bd76b64097b86828a286bc6, type: 3}
       propertyPath: m_Name
       value: Air Transport
+      objectReference: {fileID: 0}
+    - target: {fileID: 1182439771315072052, guid: 4d0c2e8c3bd76b64097b86828a286bc6, type: 3}
+      propertyPath: <Owner>k__BackingField
+      value: 1
       objectReference: {fileID: 0}
     - target: {fileID: 2636797422651241148, guid: 4d0c2e8c3bd76b64097b86828a286bc6, type: 3}
       propertyPath: m_LocalPosition.x
Index: Assets/Scripts/EventBus/Bus.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/EventBus/Bus.cs b/Assets/Scripts/EventBus/Bus.cs
--- a/Assets/Scripts/EventBus/Bus.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/EventBus/Bus.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -1,10 +1,42 @@
+using System;
+using System.Collections.Generic;
+using GameDevTV.RTS.Units;
+
 namespace GameDevTV.RTS.EventBus
 {
     public static class Bus<T> where T : IEvent
     {
         public delegate void Event(T args);
-        public static event Event OnEvent;
+        public static Dictionary<Owner, Event> OnEvent = new()
+        {
+            { Owner.Player1, null },
+            { Owner.AI1, null },
+            { Owner.AI2, null },
+            { Owner.AI3, null },
+            { Owner.AI4, null },
+            { Owner.AI5, null },
+            { Owner.AI6, null },
+            { Owner.AI7, null },
+            { Owner.Invalid, null },
+            { Owner.Unowned, null }
+        };
+
+        public static void Raise(Owner owner, T evt) => OnEvent[owner]?.Invoke(evt);
 
-        public static void Raise(T evt) => OnEvent?.Invoke(evt);
+        public static void RegisterForAll(Event handler)
+        {
+            foreach(Owner owner in Enum.GetValues(typeof(Owner)))
+            {
+                OnEvent[owner] += handler;
+            }
+        }
+
+        public static void UnregisterForAll(Event handler)
+        {
+            foreach(Owner owner in Enum.GetValues(typeof(Owner)))
+            {
+                OnEvent[owner] -= handler;
+            }
+        }
     }
 }
\ No newline at end of file
Index: Assets/Scripts/Player/PlayerInput.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Player/PlayerInput.cs b/Assets/Scripts/Player/PlayerInput.cs
--- a/Assets/Scripts/Player/PlayerInput.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Player/PlayerInput.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -59,20 +59,20 @@
             startingFollowOffset = cinemachineFollow.FollowOffset;
             maxRotationAmount = Mathf.Abs(cinemachineFollow.FollowOffset.z);
 
-            Bus<UnitSelectedEvent>.OnEvent += HandleUnitSelected;
-            Bus<UnitDeselectedEvent>.OnEvent += HandleUnitDeselected;
-            Bus<UnitSpawnEvent>.OnEvent += HandleUnitSpawn;
-            Bus<CommandSelectedEvent>.OnEvent += HandleActionSelected;
-            Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
+            // Bus<UnitSelectedEvent>.OnEvent += HandleUnitSelected;
+            // Bus<UnitDeselectedEvent>.OnEvent += HandleUnitDeselected;
+            // Bus<UnitSpawnEvent>.OnEvent += HandleUnitSpawn;
+            // Bus<CommandSelectedEvent>.OnEvent += HandleActionSelected;
+            // Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
         }
 
         private void OnDestroy()
         {
-            Bus<UnitSelectedEvent>.OnEvent -= HandleUnitSelected;
-            Bus<UnitDeselectedEvent>.OnEvent -= HandleUnitDeselected;
-            Bus<UnitSpawnEvent>.OnEvent -= HandleUnitSpawn;
-            Bus<CommandSelectedEvent>.OnEvent -= HandleActionSelected;
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+            // Bus<UnitSelectedEvent>.OnEvent -= HandleUnitSelected;
+            // Bus<UnitDeselectedEvent>.OnEvent -= HandleUnitDeselected;
+            // Bus<UnitSpawnEvent>.OnEvent -= HandleUnitSpawn;
+            // Bus<CommandSelectedEvent>.OnEvent -= HandleActionSelected;
+            // Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
         }
 
         private void HandleUnitSelected(UnitSelectedEvent evt)
Index: Assets/Scripts/Player/Supplies.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Player/Supplies.cs b/Assets/Scripts/Player/Supplies.cs
--- a/Assets/Scripts/Player/Supplies.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Player/Supplies.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -23,12 +23,12 @@
 
         private void Awake()
         {
-            Bus<SupplyEvent>.OnEvent += HandleSupplyEvent;
+            // Bus<SupplyEvent>.OnEvent += HandleSupplyEvent;
         }
 
         private void OnDestroy()
         {
-            Bus<SupplyEvent>.OnEvent -= HandleSupplyEvent;
+            // Bus<SupplyEvent>.OnEvent -= HandleSupplyEvent;
         }
 
         private void HandleSupplyEvent(SupplyEvent evt)
Index: Assets/Scripts/UI/Containers/ActionsUI.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/UI/Containers/ActionsUI.cs b/Assets/Scripts/UI/Containers/ActionsUI.cs
--- a/Assets/Scripts/UI/Containers/ActionsUI.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/UI/Containers/ActionsUI.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -58,7 +58,8 @@
 
         private UnityAction HandleClick(BaseCommand action)
         {
-            return () => Bus<CommandSelectedEvent>.Raise(new CommandSelectedEvent(action));
+            return null;
+            // return () => Bus<CommandSelectedEvent>.Raise(new CommandSelectedEvent(action));
         }
     }
 }
Index: Assets/Scripts/UI/Containers/BuildingSelectedUI.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/UI/Containers/BuildingSelectedUI.cs b/Assets/Scripts/UI/Containers/BuildingSelectedUI.cs
--- a/Assets/Scripts/UI/Containers/BuildingSelectedUI.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/UI/Containers/BuildingSelectedUI.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -29,7 +29,7 @@
                 buildingUnderConstructionUI.EnableFor(building);
                 buildingBuildingUI.Disable();
                 singleUnitSelectedUI.Disable();
-                Bus<BuildingSpawnEvent>.OnEvent += HandleBuildingSpawn;
+                // Bus<BuildingSpawnEvent>.OnEvent += HandleBuildingSpawn;
             }
         }
 
@@ -38,7 +38,7 @@
             buildingBuildingUI.Disable();
             singleUnitSelectedUI.Disable();
             buildingUnderConstructionUI.Disable();
-            Bus<BuildingSpawnEvent>.OnEvent -= HandleBuildingSpawn;
+            // Bus<BuildingSpawnEvent>.OnEvent -= HandleBuildingSpawn;
             if (selectedBuilding != null)
             {
                 selectedBuilding.OnQueueUpdated -= OnBuildingQueueUpdated;
@@ -64,7 +64,7 @@
         {
             if (evt.Building == selectedBuilding)
             {
-                Bus<BuildingSpawnEvent>.OnEvent -= HandleBuildingSpawn;
+                // Bus<BuildingSpawnEvent>.OnEvent -= HandleBuildingSpawn;
                 OnBuildingQueueUpdated();
                 buildingUnderConstructionUI.Disable();
             }
Index: Assets/Scripts/UI/RuntimeUI.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/UI/RuntimeUI.cs b/Assets/Scripts/UI/RuntimeUI.cs
--- a/Assets/Scripts/UI/RuntimeUI.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/UI/RuntimeUI.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -20,12 +20,12 @@
 
         private void Awake()
         {
-            Bus<UnitSelectedEvent>.OnEvent += HandleUnitSelected;
-            Bus<UnitDeselectedEvent>.OnEvent += HandleUnitDeselected;
-            Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
-            Bus<SupplyEvent>.OnEvent += HandleSupplyChange;
-            Bus<UnitLoadEvent>.OnEvent += HandleLoadUnit;
-            Bus<UnitUnloadEvent>.OnEvent += HandleUnloadUnit;
+            // Bus<UnitSelectedEvent>.OnEvent += HandleUnitSelected;
+            // Bus<UnitDeselectedEvent>.OnEvent += HandleUnitDeselected;
+            // Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
+            // Bus<SupplyEvent>.OnEvent += HandleSupplyChange;
+            // Bus<UnitLoadEvent>.OnEvent += HandleLoadUnit;
+            // Bus<UnitUnloadEvent>.OnEvent += HandleUnloadUnit;
         }
 
         private void Start()
@@ -39,12 +39,12 @@
 
         private void OnDestroy()
         {
-            Bus<UnitSelectedEvent>.OnEvent -= HandleUnitSelected;
-            Bus<UnitDeselectedEvent>.OnEvent -= HandleUnitDeselected;
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
-            Bus<SupplyEvent>.OnEvent -= HandleSupplyChange;
-            Bus<UnitLoadEvent>.OnEvent -= HandleLoadUnit;
-            Bus<UnitUnloadEvent>.OnEvent -= HandleUnloadUnit;
+            // Bus<UnitSelectedEvent>.OnEvent -= HandleUnitSelected;
+            // Bus<UnitDeselectedEvent>.OnEvent -= HandleUnitDeselected;
+            // Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+            // Bus<SupplyEvent>.OnEvent -= HandleSupplyChange;
+            // Bus<UnitLoadEvent>.OnEvent -= HandleLoadUnit;
+            // Bus<UnitUnloadEvent>.OnEvent -= HandleUnloadUnit;
         }
 
         private void HandleUnitSelected(UnitSelectedEvent evt)
Index: Assets/Scripts/Units/AbstractCommandable.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/AbstractCommandable.cs b/Assets/Scripts/Units/AbstractCommandable.cs
--- a/Assets/Scripts/Units/AbstractCommandable.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/AbstractCommandable.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -35,7 +35,7 @@
             }
 
             IsSelected = true;
-            Bus<UnitSelectedEvent>.Raise(new UnitSelectedEvent(this));
+            Bus<UnitSelectedEvent>.Raise(Owner, new UnitSelectedEvent(this));
         }
 
         public virtual void Deselect()
@@ -48,7 +48,7 @@
             IsSelected = false;
             SetCommandOverrides(null);
 
-            Bus<UnitDeselectedEvent>.Raise(new UnitDeselectedEvent(this));
+            Bus<UnitDeselectedEvent>.Raise(Owner, new UnitDeselectedEvent(this));
         }
 
         public void SetCommandOverrides(BaseCommand[] commands)
@@ -64,7 +64,7 @@
 
             if (IsSelected)
             {
-                Bus<UnitSelectedEvent>.Raise(new UnitSelectedEvent(this));
+                Bus<UnitSelectedEvent>.Raise(Owner, new UnitSelectedEvent(this));
             }
         }
 
Index: Assets/Scripts/Units/AbstractUnit.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/AbstractUnit.cs b/Assets/Scripts/Units/AbstractUnit.cs
--- a/Assets/Scripts/Units/AbstractUnit.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/AbstractUnit.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -35,7 +35,7 @@
             base.Start();
             CurrentHealth = UnitSO.Health;
             MaxHealth = UnitSO.Health;
-            Bus<UnitSpawnEvent>.Raise(new UnitSpawnEvent(this));
+            Bus<UnitSpawnEvent>.Raise(Owner, new UnitSpawnEvent(this));
 
             if (DamageableSensor != null)
             {
@@ -120,7 +120,7 @@
 
         protected virtual void OnDestroy()
         {
-            Bus<UnitDeathEvent>.Raise(new UnitDeathEvent(this));
+            Bus<UnitDeathEvent>.Raise(Owner, new UnitDeathEvent(this));
         }
     }
 }
Index: Assets/Scripts/Units/AirTransport.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/AirTransport.cs b/Assets/Scripts/Units/AirTransport.cs
--- a/Assets/Scripts/Units/AirTransport.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/AirTransport.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -65,7 +65,7 @@
                 }
 
                 loadedUnits.Remove(unit);
-                Bus<UnitUnloadEvent>.Raise(new UnitUnloadEvent(unit, this));
+                Bus<UnitUnloadEvent>.Raise(Owner, new UnitUnloadEvent(unit, this));
                 return true;
             }
 
@@ -100,7 +100,7 @@
             UsedCapacity += transportable.TransportCapacityUsage;
 
             loadedUnits.Add(transportable);
-            Bus<UnitLoadEvent>.Raise(new UnitLoadEvent(transportable, this));
+            Bus<UnitLoadEvent>.Raise(Owner, new UnitLoadEvent(transportable, this));
 
             if(graphAgent.GetVariable("LoadUnitTargets", out BlackboardVariable<List<GameObject>> loadUnitsVariable))
             {
Index: Assets/Scripts/Units/BaseBuilding.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/BaseBuilding.cs b/Assets/Scripts/Units/BaseBuilding.cs
--- a/Assets/Scripts/Units/BaseBuilding.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/BaseBuilding.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -44,8 +44,8 @@
             }
             Progress = new BuildingProgress(BuildingProgress.BuildingState.Completed, Progress.StartTime, 1);
             unitBuildingThis = null;
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
-            Bus<BuildingSpawnEvent>.Raise(new BuildingSpawnEvent(this));
+            Bus<UnitDeathEvent>.OnEvent[Owner] -= HandleUnitDeath;
+            Bus<BuildingSpawnEvent>.Raise(Owner, new BuildingSpawnEvent(this));
         }
 
         public void BuildUnit(AbstractUnitSO unit)
@@ -56,8 +56,8 @@
                 return;
             }
 
-            Bus<SupplyEvent>.Raise(new SupplyEvent(-unit.Cost.Minerals, unit.Cost.MineralsSO));
-            Bus<SupplyEvent>.Raise(new SupplyEvent(-unit.Cost.Gas, unit.Cost.GasSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(-unit.Cost.Minerals, unit.Cost.MineralsSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(-unit.Cost.Gas, unit.Cost.GasSO));
 
             buildingQueue.Add(unit);
             if (buildingQueue.Count == 1)
@@ -79,8 +79,8 @@
             }
 
             AbstractUnitSO unitSO = buildingQueue[index];
-            Bus<SupplyEvent>.Raise(new SupplyEvent(unitSO.Cost.Minerals, unitSO.Cost.MineralsSO));
-            Bus<SupplyEvent>.Raise(new SupplyEvent(unitSO.Cost.Gas, unitSO.Cost.GasSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(unitSO.Cost.Minerals, unitSO.Cost.MineralsSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(unitSO.Cost.Gas, unitSO.Cost.GasSO));
             buildingQueue.RemoveAt(index);
             if (index == 0)
             {
@@ -119,8 +119,8 @@
                 Heal(1);
             }
 
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
-            Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
+            Bus<UnitDeathEvent>.OnEvent[Owner] -= HandleUnitDeath;
+            Bus<UnitDeathEvent>.OnEvent[Owner] += HandleUnitDeath;
         }
 
         private void HandleUnitDeath(UnitDeathEvent evt)
@@ -133,7 +133,7 @@
                     (Time.time - Progress.StartTime) / BuildingSO.BuildTime
                 );
 
-                Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+                Bus<UnitDeathEvent>.OnEvent[Owner] -= HandleUnitDeath;
             }
         }
 
@@ -160,7 +160,7 @@
 
         private void OnDestroy()
         {
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+            Bus<UnitDeathEvent>.OnEvent[Owner] -= HandleUnitDeath;
         }
     }
 }
Index: Assets/Scripts/Units/DamageableSensor.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/DamageableSensor.cs b/Assets/Scripts/Units/DamageableSensor.cs
--- a/Assets/Scripts/Units/DamageableSensor.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/DamageableSensor.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -34,7 +34,7 @@
 
             if (damageables.Count == 1)
             {
-                Bus<UnitDeathEvent>.OnEvent += HandleUnitDeath;
+                Bus<UnitDeathEvent>.RegisterForAll(HandleUnitDeath);
             }
         }
 
@@ -47,18 +47,18 @@
 
             if (damageables.Count == 0)
             {
-                Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+                Bus<UnitDeathEvent>.UnregisterForAll(HandleUnitDeath);
             }
         }
 
         private void OnDestroy()
         {
-            Bus<UnitDeathEvent>.OnEvent -= HandleUnitDeath;
+            Bus<UnitDeathEvent>.UnregisterForAll(HandleUnitDeath);
         }
 
         private void HandleUnitDeath(UnitDeathEvent evt)
         {
-            if (damageables.Remove(evt.Unit))
+            if (damageables.Contains(evt.Unit))
             {
                 OnTriggerExit(evt.Unit.GetComponent<Collider>());
             }
Index: Assets/Scripts/Units/Worker.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Assets/Scripts/Units/Worker.cs b/Assets/Scripts/Units/Worker.cs
--- a/Assets/Scripts/Units/Worker.cs	(revision e3ea8e11289eec691aeec29b5500718851cd371f)
+++ b/Assets/Scripts/Units/Worker.cs	(revision 7889c5b6fc3bd230eaf41981f4036c4171bec113)
@@ -74,8 +74,8 @@
             graphAgent.SetVariableValue("Command", UnitCommands.BuildBuilding);
 
             SetCommandOverrides(new BaseCommand[] { CancelBuildingCommand });
-            Bus<SupplyEvent>.Raise(new SupplyEvent(-building.Cost.Minerals, building.Cost.MineralsSO));
-            Bus<SupplyEvent>.Raise(new SupplyEvent(-building.Cost.Gas, building.Cost.GasSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(-building.Cost.Minerals, building.Cost.MineralsSO));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(-building.Cost.Gas, building.Cost.GasSO));
 
             return instance;
         }
@@ -102,11 +102,11 @@
                 Destroy(buildingVariable.Value.gameObject);
 
                 BuildingSO buildingSO = buildingVariable.Value.BuildingSO;
-                Bus<SupplyEvent>.Raise(new SupplyEvent(
+                Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(
                     Mathf.FloorToInt(0.75f * buildingSO.Cost.Minerals),
                     buildingSO.Cost.MineralsSO
                 ));
-                Bus<SupplyEvent>.Raise(new SupplyEvent(
+                Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(
                     Mathf.FloorToInt(0.75f * buildingSO.Cost.Gas),
                     buildingSO.Cost.GasSO
                 ));
@@ -129,12 +129,12 @@
                 SetCommandOverrides(null);
             }
 
-            Bus<UnitDeselectedEvent>.Raise(new UnitDeselectedEvent(this));
+            Bus<UnitDeselectedEvent>.Raise(Owner, new UnitDeselectedEvent(this));
         }
 
         private void HandleGatherSupplies(GameObject self, int amount, SupplySO supply)
         {
-            Bus<SupplyEvent>.Raise(new SupplyEvent(amount, supply));
+            Bus<SupplyEvent>.Raise(Owner, new SupplyEvent(amount, supply));
         }
 
         private void HandleBuildingEvent(GameObject self, BuildingEventType eventType, BaseBuilding building)
